### PR TITLE
PERF: various perf improvements of chat-live-pane

### PR DESCRIPTION
**Pull request by jjaffeux**

### CODE REVIEW

File: plugins/chat/assets/javascripts/discourse/components/chat-composer-uploads.js [hunk 1]
```diff
@@ -26,12 +26,14 @@ export default Component.extend(UppyUploadMixin, {
 
   didReceiveAttrs() {
     this._super(...arguments);
+    if (this._uppyInstance?.getState?.()?.totalProgress > 0) {
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/components/chat-composer-uploads.js:29]: This is a micro optim, but it avoids doing all kind of unecessary work like sending an app event

File: plugins/chat/assets/javascripts/discourse/components/chat-live-pane.js [hunk 2]
```diff
@@ -472,25 +473,25 @@ export default class ChatLivePane extends Component {
...
-      if (this.isTowardsTop || this.isAtTop) {
+      if (this.isAtTop) {
         this.fetchMoreMessages({ direction: PAST });
       }
     } else {
-      if (this.isTowardsBottom || this.isAtBottom) {
+      if (this.isAtBottom) {
         this.fetchMoreMessages({ direction: FUTURE });
       }
     }
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/components/chat-live-pane.js:491]: Changed the logic from % based to pixel length. The % solution was less accurate when there was a lot of messages loaded as 5% could actually be a lot of messages.

File: plugins/chat/assets/javascripts/discourse/components/chat-live-pane.js [hunk 3]
```diff
@@ -472,25 +473,25 @@ export default class ChatLivePane extends Component {
...
 
     this._previousScrollTop = scrollPosition;
-    this.onScrollEndedHandler = discourseLater(this, this.onScrollEnded, 25);
+    this.onScrollEndedHandler = discourseLater(this, this.onScrollEnded, 150);
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/components/chat-live-pane.js:494]: Leaves more time to set `isScrolling` to false, this is done to limit cases were we try to load more and the messages have not yet be appended to container and scroll bar has not moved back to middle

File: plugins/chat/assets/javascripts/discourse/components/chat-message-in-reply-to-indicator.js [hunk 4]
```diff
@@ -14,7 +14,10 @@ export default class ChatMessageInReplyToIndicator extends Component {
 
   get model() {
     if (this.hasThread) {
-      return [this.args.message.threadId];
+      return [
+        ...this.args.message.channel.routeModels,
+        this.args.message.threadId,
+      ];
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/components/chat-message-in-reply-to-indicator.js:20]: Simple bug with threads which were not having the correct params

File: plugins/chat/assets/javascripts/discourse/components/chat-message.hbs [hunk 5]
```diff
@@ -46,6 +46,7 @@
   {{on "touchstart" this.handleTouchStart passive=true}}
   {{on "touchend" this.handleTouchEnd passive=true}}
   {{on "mouseenter" (fn @onHoverMessage @message (hash desktopOnly=true))}}
+  {{on "mousemove" (fn @onHoverMessage @message (hash desktopOnly=true))}}
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/components/chat-message.hbs:49]: This is a consequence of the longer delay on scroll ended. Without this change your mouse could end up on a message at the end of the scroll and we would still not consider the scroll as ended and wouldnt trigger hovering (which is the whole of this optimization, not trigger mouseover during scroll as it's expensive). Now as soon as you move the mouse we try to hover again.

File: plugins/chat/assets/javascripts/discourse/components/full-page-chat.hbs [hunk 6]
```diff
@@ -1,6 +1,6 @@
-{{#if this.chat.activeChannel}}
+{{#if @channel.id}}
   <ChatLivePane
-    @channel={{this.chat.activeChannel}}
+    @channel={{@channel}}
     @targetMessageId={{readonly @targetMessageId}}
   />
 {{/if}}
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/components/full-page-chat.hbs:6]: minor optim to rely on route params

File: plugins/chat/assets/javascripts/discourse/routes/chat-channel.js [hunk 7]
```diff
@@ -13,6 +13,16 @@ export default class ChatChannelRoute extends DiscourseRoute {
...
+    if (transition?.to?.name === "chat.channel.index") {
+      const targetChannelId = transition?.to?.parent?.params?.channelId;
+      if (
+        targetChannelId &&
+        parseInt(targetChannelId, 10) !== this.chat.activeChannel.id
+      ) {
+        this.chat.activeChannel.clearMessages();
+      }
+    }
+
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/routes/chat-channel.js:25]: always clear messages if we transition to a different channel

File: plugins/chat/assets/javascripts/discourse/templates/chat-channel-near-message.hbs [hunk 8]
```diff
@@ -1 +0,0 @@
-<FullPageChat />
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/templates/chat-channel-near-message.hbs:1]: this is never used

**Review by ZogStriP — APPROVED** [review]

**Pull request merged**